### PR TITLE
Implement partitions over Domains.

### LINKED ISSUE #1049: Expand `Domain` functionality
To further improve `GQCP`, we should extend the `Domain` functionality that allows for the partitioning of electron numbers, orbitals, and ONVs into distinct domains. This marks the second step of the refactoring process from `GQCD` to `GQCP`.

### PR DESCRIPTION
**Pull request by dariavh**
**Short description**

This PR will provide new features regarding the partition of domains, electron numbers and orbitals over the `Domain`s defined by `GQCP`.

**Related issues**
Closes #1049.

**To do**

- [x] (if applicable, when creating new source files): don't forget to update the Doxygen file, the collective `gqcp.hpp` include header and the CMake files
- [x] Implement `SimplePartition` as a base class for all partitions.
- [x] Implement `DomainPartition` as a base class for partitions of domains.
- [x] Implement `DiscreteDomainPartition` that divides a discrete set of orbitals into domains.
- [x] Generalize functionality `overlapWithONV` from `HubbardDomain` to `DiscreteDomain`.
- [x] Implement `ONVPartition` to create new ONVs for each (discrete) domain separately.
- [x] Implement `ElectronPartition` (spin-unresolved and spin-resolved) to specify the electron occupation in each domain.
- [x] Write tests.
- [x] Write Python bindings.

### CODE REVIEW

**Comment by dariavh** [comment]
@xdvriend This PR is quite a bulky one because I introduced a lot of new `Partition`-related implementations. I have rebased this branch with `develop` to allow for working Docker containers again, but this rebase added a lot of files that do not need your review again. The files unique to this PR are in the `Domain` and `Partition` folder (`hpp`, `cpp`, `test` and `bindings`). I have also added the calculation of probabilities of finding a given distribution of electrons over a set of domains to `LinearExpansion`.

File: gqcp/include/Partition/ONVPartition.hpp [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    }
+
+    /**
+     *  @return The phase factor of the ONV partition.
+     */
+    int phaseFactor() const { return this->phase_factor; }
+};
+
+
+}  // namespace GQCP
```
Comment (xdvriend) [inline gqcp/include/Partition/ONVPartition.hpp]:
```suggestion
}  // namespace GQCP

```

File: gqcp/include/Partition/SimplePartition.hpp [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+/**
+ *  A type specifically designed to act as a parent class for e.g. domain-like, electron-like or orbital-like partitions in order to share common functionality.
+ *
+ *  @tparam _DerivedPartition              The type of the partition that derives from this class, enabling CRTP and compile-time polymorphism.
+ */
+template <typename _DerivedPartition>
+class SimplePartition:
+    public CRTP<_DerivedPartition> {
+public:
+    // The type of partition that derives from this class, enabling CRTP and compile-time polymorphism..
```
Comment (xdvriend) [inline gqcp/include/Partition/SimplePartition.hpp]:
```suggestion
    // The type of partition that derives from this class, enabling CRTP and compile-time polymorphism.
```

File: gqcp/include/Partition/SimplePartition.hpp [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+
+public:
+    /*
+     *  MARK: Constructors
+     */
+
+    SimplePartition() {}
+
+    // COPY CONSTRUCTOR
```
Comment (xdvriend) [inline gqcp/include/Partition/SimplePartition.hpp]: Maybe adjust the documentation here to be uniform with the rest of the constructor headers.

File: gqcp/include/Partition/SimplePartition.hpp [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+     */
+    virtual std::string asString() const = 0;
+
+    /**
+     *  @return     The dimension of this partition.
+     */
+    size_t dimension() const { return this->partition.size(); }
+
+    /**
+     *  @param other        the other partition
```
Comment (xdvriend) [inline gqcp/include/Partition/SimplePartition.hpp]:
```suggestion
     *  @param other        The other partition.
```

File: gqcp/include/Partition/SimplePartition.hpp [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        for (size_t i = 0; i < this->dimension(); i++) {
+            if (this->operator()(i) != other(i)) {
+                equal = false;
+            }
+        }
+        return equal;
+    }
+
+    /**
+     *  @param other        the other partition
```
Comment (xdvriend) [inline gqcp/include/Partition/SimplePartition.hpp]:
```suggestion
     *  @param other        The other partition.
```

File: gqcp/src/Partition/SpinResolvedElectronPartition.cpp [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+/**
+ *  @return     The number of spin-resolved electrons the partition contains.
+ */
+size_t SpinResolvedElectronPartition::numberOfElectrons() const {
+    return this->alpha().numberOfElectrons() + this->beta().numberOfElectrons();
+}
+
+
+/**
+ *  @param other        the other partition
```
Comment (xdvriend) [inline gqcp/src/Partition/SpinResolvedElectronPartition.cpp]:
```suggestion
 *  @param other        The other partition.
```

File: gqcp/src/Partition/SpinResolvedElectronPartition.cpp [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        if (alpha_partition(i) + beta_partition(i) != other(i)) {
+            equal = false;
+        }
+    }
+    return equal;
+}
+
+
+/**
+ *  @param other        the other partition
```
Comment (xdvriend) [inline gqcp/src/Partition/SpinResolvedElectronPartition.cpp]:
```suggestion
 *  @param other        The other partition.
```

**Review by xdvriend — APPROVED** [review]
Some minor comments, but code related everything looks great!

**Pull request merged**